### PR TITLE
[network] Improved Connection Manager

### PR DESCRIPTION
**Pull request by 0xcb9ff9**
# Description

Issues:

- identity protocol stream is never closed, `IdentityService` cannot save the identity protocol stream during the handshake.
- if `bootnode` shutdown (like network interruption), never re-connect known peers
- `PeerConnInfo` not match the libp2p connection

This PR improved connection manager, key point:

- wrap grpc client, use runtime.SetFinalizer cleanup
- cleanup connect by background task
- save known peer address info, random re-connect

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels

## Testing

- [x] I have tested this code with the official test suite

### CODE REVIEW

File: network/discovery/discovery.go [hunk 1]
```diff
@@ -299,7 +298,7 @@ func (d *DiscoveryService) findPeersCall(
 
 	clt, clientErr := d.baseServer.NewDiscoveryClient(peerID)
 	if clientErr != nil {
-		return nil, fmt.Errorf("unable to create new discovery client connection, %w", clientErr)
+		return nil, clientErr
```
Comment (DarianShawn) [inline network/discovery/discovery.go:385]: Why stop wrapping error with the context?
The previous error is more readable.

File: network/server_discovery.go [hunk 2]
```diff
@@ -25,7 +25,8 @@ func (s *DefaultServer) GetRandomBootnode() *peer.AddrInfo {
 	nonConnectedNodes := make([]*peer.AddrInfo, 0)
 
 	for _, v := range s.bootnodes.getBootnodes() {
-		if !s.HasPeer(v.ID) {
+		// don't use HasPeer()
+		if !s.IsConnected(v.ID) && v.ID != s.host.ID() {
```
Comment (DarianShawn) [inline network/server_discovery.go]: Any more comment for this modification?
I don't see any difference between `s.IsConnected` and `s.HasPeer`.

File: network/discovery/discovery.go (hunk 1 above)
Comment (0xcb9ff9) [reply network/discovery/discovery.go:385]: [discovery.go#L383](https://github.com/dogechain-lab/dogechain/blob/b2cfd24ba7d31b6c008a2fa972510074337fbcde/network/discovery/discovery.go#L383) and [discovery.go#L435](https://github.com/dogechain-lab/dogechain/blob/b2cfd24ba7d31b6c008a2fa972510074337fbcde/network/discovery/discovery.go#L435), print error is too long,  [server_discovery.go#L53](https://github.com/dogechain-lab/dogechain/blob/b2cfd24ba7d31b6c008a2fa972510074337fbcde/network/server_discovery.go#L53) exist are duplicate descriptions

File: network/server_discovery.go (hunk 2 above)
Comment (0xcb9ff9) [reply network/server_discovery.go]: `IsConnected` check libp2p connection, `HasPeer` check `PeerConnInfo` is exist, `PeerConnInfo` exists but the `connection` does not necessarily exist (this is bug,I need more time to fix it.)

File: network/wrappers/interface.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+var (
+	ErrClientClosed = errors.New("client is closed")
+)
+
+const (
+	errClosedClientInFinalizer = "closing client connection in finalizer"
+)
+
+type GrpcClientWrapper interface {
```
Comment (0xcb9ff9) [inline network/wrappers/interface.go]: Wrapping grpc connections, avoid package cyclic references.

File: txpool/event_subscription_test.go [hunk 4]
```diff
@@ -146,8 +146,12 @@ func TestEventSubscription_ProcessedEvents(t *testing.T) {
 			}
 
 			wg.Wait()
-			eventWaitCtx, eventWaitFn := context.WithTimeout(context.Background(), time.Second*10)
```
Comment (0xcb9ff9) [inline txpool/event_subscription_test.go:149]: 10 seconds is too short, needs to be calculated based on the size of the event

File: txpool/event_subscription_test.go [hunk 5]
```diff
@@ -146,8 +146,12 @@ func TestEventSubscription_ProcessedEvents(t *testing.T) {
...
+
+			eventWaitCtx, eventWaitFn := context.WithTimeout(
+				context.Background(),
+				time.Second*time.Duration(testCase.expectedProcessed+1))
```
Comment (0xcb9ff9) [inline txpool/event_subscription_test.go]: 10 seconds is too short, needs to be calculated based on the size of the event

File: network/wrappers/discovery.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
+//nolint:dupl
+package wrappers
```
Comment (DarianShawn) [inline network/wrappers/discovery.go]: `wrappers` is vague. Name it to `client` for a clearer meaning.

File: network/discovery/discovery.go [hunk 7]
```diff
@@ -87,6 +88,74 @@ type networkingServer interface {
 	PeerCount() int64
 }
 
+// peerAddreStore is a struct that contains the peer address information
+type peerAddreStore struct {
+	peerAddressLock sync.RWMutex               // protects the peerAddress map
```
Comment (DarianShawn) [inline network/discovery/discovery.go]: `lock` is enough for the name.

File: network/discovery/discovery.go [hunk 8]
```diff
@@ -87,6 +88,74 @@ type networkingServer interface {
...
+	peerAddress     map[peer.ID]*peer.AddrInfo // stores the peer address information
```
Comment (DarianShawn) [inline network/discovery/discovery.go]: `peerAddresses`?

File: network/wrappers/interface.go (hunk 3 above)
Comment (DarianShawn) [inline network/wrappers/interface.go]: How about `GrpcClientCloser`?

File: network/server.go [hunk 9]
```diff
@@ -217,33 +228,64 @@ func (s *DefaultServer) HasFreeConnectionSlot(direction network.Direction) bool
...
+	pci.connDirections[direction] = true
+}
+
+// removeConnDirection adds a connection direction
+func (pci *PeerConnInfo) removeConnDirection(direction network.Direction) {
+	pci.connDirections[direction] = false
+}
+
+// getConnDirection returns the connection direction
+func (pci *PeerConnInfo) getConnDirection(direction network.Direction) bool {
```
Comment (DarianShawn) [inline network/server.go]: How about rename it `existsConnDirection`?

File: network/server.go [hunk 10]
```diff
@@ -217,33 +228,64 @@ func (s *DefaultServer) HasFreeConnectionSlot(direction network.Direction) bool
...
+	exist, ok := pci.connDirections[direction]
+	if !ok {
+		return false
+	}
+
+	return exist
+}
+
+func (pci *PeerConnInfo) noConnectionAvailable() bool {
```
Comment (DarianShawn) [inline network/server.go:256]: `connectionNotAvailable` is more intuitive.

File: network/server.go [hunk 11]
```diff
@@ -391,7 +435,8 @@ func (s *DefaultServer) keepAliveStaticPeerConnections() {
 				allConnected = false
 			}
 
-			s.joinPeer(add)
+			s.logger.Info("reconnect static peer", "addr", add.String())
+			s.addToDialQueue(add, common.PriorityRequestedDial)
```
Comment (DarianShawn) [inline network/server.go:442]: What event is emitted when a peer is reconnected?

File: network/server.go [hunk 12]
```diff
@@ -433,46 +478,181 @@ func (s *DefaultServer) setupBootnodes() error {
...
 		}
 
-		if s.PeerCount() >= MinimumPeerConnections {
+		delay.Reset(DefaultKeepAliveTimer)
+
+		// if discovery service is not available, wait for it to be available
+		// this happens when the startup is not complete
+		if s.discovery == nil {
+			s.logger.Error("discovery service is nil")
+			delay.Reset(DefaultKeepAliveTimer * 12) // wait 2 minutes before trying again
```
Comment (DarianShawn) [inline network/server.go]: I would suggest a new constant instead of doing magic math here.

File: network/server.go [hunk 13]
```diff
@@ -433,46 +478,181 @@ func (s *DefaultServer) setupBootnodes() error {
...
-			if randPeer != nil && !s.IsConnected(*randPeer) {
-				s.addToDialQueue(s.GetPeerInfo(*randPeer), common.PriorityRandomDial)
+		// first get all the peers from the peerstore
+		// mark connection if not create PeerConnInfo or pending status
+		// call DisconnectFromPeer to clean up the connection marked in the last round
+
+		// use a waitgroup avoid disconnect deadlock
+		var waitingPeersDiscovered sync.WaitGroup
+
+		disconnectFlag := false
```
Comment (DarianShawn) [inline network/server.go]: This variable is useless, `waitingPeersDiscovered` already does everything you need.

File: network/server.go [hunk 14]
```diff
@@ -433,46 +478,181 @@ func (s *DefaultServer) setupBootnodes() error {
...
-			s.addToDialQueue(randomNode, common.PriorityRandomDial)
 		}
+
+		pendingConnectMark = copyMark
+
+		// next check libp2p connections,
+		// if connect count is enough, wait next loop
+
+		if len(peers) >= maxPeers {
+			delay.Reset(DefaultKeepAliveTimer * 6) // wait 1 minute before trying again
```
Comment (DarianShawn) [inline network/server.go]: Magic math again. Use constant plz.

File: network/server.go [hunk 15]
```diff
@@ -433,46 +478,181 @@ func (s *DefaultServer) setupBootnodes() error {
...
 
 	return nil
 }
 
-// keepAliveMinimumPeerConnections will attempt to make new connections
+// keepAvailablePeerConnections clear the peerstore of peers that are not connected
+// and will attempt to make new connections
 // if the active peer count is lesser than the specified limit.
-func (s *DefaultServer) keepAliveMinimumPeerConnections() {
+func (s *DefaultServer) keepAvailablePeerConnections() {
```
Comment (DarianShawn) [inline network/server.go]: TL;DR;
It is full of `for` loop, `if` indentation, hard to read and understand.
For the sake of each contributor, please try to break this down into some smaller methods.

File: network/server.go [hunk 16]
```diff
@@ -844,38 +973,45 @@ func (s *DefaultServer) SaveProtocolStream(
...
+	s.logger.Debug("create grpc client", "protocol", protocol, "peer", peerID)
+
+	// don't need context.WithTimeout, rpc client is fake
 	return p.Client(context.Background(), stream), nil
 }
 
 func (s *DefaultServer) NewStream(proto string, id peer.ID) (network.Stream, error) {
-	return s.host.NewStream(context.Background(), id, protocol.ID(proto))
+	// don't dial
+	ctx := network.WithNoDial(context.Background(), "grpc client stream")
```
Comment (DarianShawn) [inline network/server.go:842]: Could you add some more comments on why we should not dial here, and an additional comment on the method for any callers to notice?

File: network/wrappers/identity.go [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/hashicorp/go-hclog"
+	"go.uber.org/atomic"
+
+	rawGrpc "google.golang.org/grpc"
+)
+
+type IdentityClient interface {
+	Hello(ctx context.Context, in *proto.Status) (*proto.Status, error)
+
+	GrpcClientWrapper
```
Comment (DarianShawn) [inline network/wrappers/identity.go]: It is common to put embedded fields first.

File: txpool/txpool.go [hunk 18]
```diff
@@ -337,7 +337,7 @@ func (p *TxPool) Start() {
 
 // Close shuts down the pool's main loop.
 func (p *TxPool) Close() {
-	if !p.isClosed.CAS(false, true) {
+	if !p.isClosed.CompareAndSwap(false, true) {
```
Comment (DarianShawn) [inline txpool/txpool.go:341]: Another deprecated `CAS` call in the `protocol/client.go`.

**Review by DarianShawn — COMMENTED** [review]
Thanks for the great job. 🙌
Some questions here though.

File: network/server.go (hunk 15 above)
Comment (0xcb9ff9) [reply network/server.go]: [eb532cb](https://github.com/dogechain-lab/dogechain/pull/305/commits/eb532cb236650e0da40861f7cff030e48cfdd2d9) use FSM replace

File: network/server.go (hunk 16 above)
Comment (0xcb9ff9) [reply network/server.go:842]: add it [f10b254](https://github.com/dogechain-lab/dogechain/pull/305/commits/f10b25407701e7017a7006b6c932c2b6967ffb20)

File: network/wrappers/identity.go (hunk 17 above)
Comment (0xcb9ff9) [reply network/wrappers/identity.go]: done

File: txpool/txpool.go (hunk 18 above)
Comment (0xcb9ff9) [reply txpool/txpool.go:341]: done

File: network/identity/identity.go [hunk 19]
```diff
@@ -126,32 +144,34 @@ func (i *IdentityService) GetNotifyBundle() *network.NotifyBundle {
...
 
 // addPendingStatus adds the pending status to a peer,
-// and updates adequate counter information [Thread safe]
-func (i *IdentityService) addPendingStatus(id peer.ID, direction network.Direction) {
-	if _, loaded := i.pendingPeerConnections.LoadOrStore(id, direction); !loaded {
-		i.baseServer.UpdatePendingConnCount(1, direction)
-	}
+// and updates adequate counter information
+func (i *IdentityService) addPendingStatus(peerID peer.ID, direction network.Direction) {
+	i.pendingPeerConnections[peerID] = struct{}{}
```
Comment (DarianShawn) [inline network/identity/identity.go:173]: Do we need lock here?

File: network/keep_available.go [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+
+		switch ka.status {
+		case kasWeakUp:
+			// first check discovery service is ready
+			ka.checkDiscoveryServiceReady()
+		case kasDiscoveryWaiting:
+			ka.timer.Reset(waitDiscoverDuration)
+
+			return
+		case kasMarkConnectionPanding:
```
Comment (DarianShawn) [inline network/keep_available.go]: typo here.
`Pending`

File: network/keep_available.go [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+			ka.markPendingConnection()
```
Comment (DarianShawn) [inline network/keep_available.go:163]: Are those methods reentrant?
E.g. the timer trigger it again when it is not done yet.

**Review by DarianShawn — COMMENTED** [review]
Some issues left

File: network/identity/identity.go (hunk 19 above)
Comment (0xcb9ff9) [reply network/identity/identity.go:173]: no need, write lock in here [network/identity/identity.go#L10](https://github.com/dogechain-lab/dogechain/blob/bb2532215f0bd7a9b80468a764c6b20097eb9cca/network/identity/identity.go#L100)

File: network/keep_available.go (hunk 21 above)
Comment (0xcb9ff9) [reply network/keep_available.go:163]: never reentrant, this is single goroutine, Timer trigger only one pass

**Review by DarianShawn — APPROVED** [review]
LGTM 💯
Thanks for the contribution.

**Pull request merged**